File: genesis/block/block.go
package block

import (
	"crypto/sha256"
	"encoding/hex"
	"time"
)

// Block represents a single block in the chain
type Block struct {
	Index        int
	Timestamp    string
	Data         string
	PreviousHash string
	Hash         string
}

// calculateHash creates SHA256 hash of block contents
func calculateHash(b Block) string {
	record := string(rune(b.Index)) + b.Timestamp + b.Data + b.PreviousHash
	hash := sha256.Sum256([]byte(record))
	return hex.EncodeToString(hash[:])
}

// NewBlock creates a new block
func NewBlock(index int, data string, prevHash string) Block {
	block := Block{
		Index:        index,
		Timestamp:    time.Now().Format(time.RFC3339),
		Data:         data,
		PreviousHash: prevHash,
	}

	block.Hash = calculateHash(block)
	return block
}

File: src/block.rs
use sha2::{Sha256, Digest};
use chrono::Utc;

#[derive(Debug, Clone)]
pub struct Block {
    pub index: u64,
    pub timestamp: i64,
    pub data: String,
    pub previous_hash: String,
    pub nonce: u64,
    pub hash: String,
}

impl Block {
    pub fn new(index: u64, data: String, previous_hash: String) -> Self {
        let timestamp = Utc::now().timestamp();

        let mut block = Block {
            index,
            timestamp,
            data,
            previous_hash,
            nonce: 0,
            hash: String::new(),
        };

        block.hash = block.calculate_hash();
        block
    }

    pub fn calculate_hash(&self) -> String {
        let input = format!(
            "{}{}{}{}{}",
            self.index,
            self.timestamp,
            self.data,
            self.previous_hash,
            self.nonce
        );

        let mut hasher = Sha256::new();
        hasher.update(input);
        format!("{:x}", hasher.finalize())
    }
}